flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorage.java

```diff
@@ -0,0 +1,93 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+/**
+ * CheckpointStorage implements the durable storage of checkpoint data and metadata streams.
+ * An individual checkpoint or savepoint is stored to a {@link CheckpointStorageLocation},
+ * created by this class.
+ */
+public interface CheckpointStorage {
+
+	/**
+	 * Checks whether this backend supports highly available storage of data.
+	 *
+	 * <p>Some state backends may offer support for that with default settings, which makes them
+	 * suitable for zero-config prototyping, but not for actual production setups.
+	 */
+	boolean supportsHighlyAvailableStorage();
+
+	/**
+	 * Checks whether the storage has a default savepoint location configured.
+	 */
+	boolean hasDefaultSavepointLocation();
+
+	/**
+	 * Resolves the given pointer to a checkpoint/savepoint into a state handle from which the
+	 * checkpoint metadata can be read. If the state backend cannot understand the format of
+	 * the pointer (for example because it was created by a different state backend) this method
+	 * should throw an {@code IOException}.
+	 *
+	 * @param pointer The pointer to resolve.
+	 * @return The state handler from which one can read the checkpoint metadata.
+	 *
+	 * @throws IOException Thrown, if the state backend does not understand the pointer, or if
+	 *                     the pointer could not be resolved due to an I/O error.
+	 */
+	StreamStateHandle resolveCheckpoint(String pointer) throws IOException;
+
+	/**
+	 * Initializes a storage location for new checkpoint with the given ID.
+	 *
+	 * <p>The returned storage location can be used to write the checkpoint data and metadata
+	 * to and to obtain the pointers for the location(s) where the actual checkpoint data should be
+	 * stored.
+	 *
+	 * @param checkpointId The ID (logical timestamp) of the checkpoint that should be persisted.
+	 * @return A storage location for the data and metadata of the given checkpoint.
+	 *
+	 * @throws IOException Thrown if the storage location cannot be initialized due to an I/O exception.
+	 */
+	CheckpointStorageLocation initializeCheckpoint(long checkpointId) throws IOException;
```

Comment:
I would consider renaming this to something like `initializeStorageLocationForCheckpoint(...)`

Reply:
How about `initializeLocationForCheckpoint`?